File: FirebaseApp (Healthia)/app/src/main/java/com/example/firebaseapp/activitys/ChatActivity.java
package com.example.firebaseapp.activitys;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.text.format.DateFormat;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;
import com.example.firebaseapp.GetTimeAgo;
import com.example.firebaseapp.R;
import com.example.firebaseapp.adapters.AdapterChat;
import com.example.firebaseapp.models.ModelChat;
import com.example.firebaseapp.models.ModelUser;
import com.example.firebaseapp.notifications.APIService;
import com.example.firebaseapp.notifications.Client;
import com.example.firebaseapp.notifications.Data;
import com.example.firebaseapp.notifications.Response;
import com.example.firebaseapp.notifications.Sender;
import com.example.firebaseapp.notifications.Token;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.Query;
import com.google.firebase.database.ValueEventListener;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;

import retrofit2.Call;
import retrofit2.Callback;

public class ChatActivity extends AppCompatActivity {

    //views from xml
    Toolbar toolbar;
    RecyclerView recyclerView;
    ImageView profileIv;
    TextView nameTv, userStatusTv;
    EditText messageEt;
    ImageButton sendBtn;

    //firebase auth
    FirebaseAuth firebaseAuth;

    FirebaseDatabase firebaseDatabase;
    DatabaseReference usersDbRef;

    //for checking if use has seen message or not
    ValueEventListener seenListener;
    DatabaseReference userRefForSeen;

    List<ModelChat> chatList;
    AdapterChat adapterChat;

    String hisUid;
    String myUid;
    String hisImage;

    APIService apiService;
    boolean notify = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_chat);

        //init views
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        toolbar.setTitle("");
        recyclerView = findViewById(R.id.chat_recyclerView);
        profileIv = findViewById(R.id.profileIv);
        nameTv = findViewById(R.id.nameTv);
        userStatusTv = findViewById(R.id.userStatusTv);
        messageEt = findViewById(R.id.messageEt);
        sendBtn = findViewById(R.id.sendBtn);

        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        //layout (linearlayout) for recyclerview
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
        linearLayoutManager.setStackFromEnd(true);
        //recyclerview properties
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(linearLayoutManager);

        //create api service
        apiService = Client.getRetrofit("https://fcm.googleapis.com/").create(APIService.class);

        Intent intent = getIntent();
        hisUid = intent.getStringExtra("hisUid");

        //firebase auth instance
        firebaseAuth = FirebaseAuth.getInstance();

        firebaseDatabase = FirebaseDatabase.getInstance();
        usersDbRef = firebaseDatabase.getReference("Users");

        //search user to get user's info
        Query userQuery = usersDbRef.orderByChild("uid").equalTo(hisUid);

        //get user's picture and name
        userQuery.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                //check until required info is received
                for (DataSnapshot ds: snapshot.getChildren()){
                    //get data
                    String name = ""+ ds.child("name").getValue();
                    hisImage = ""+ ds.child("image").getValue();
                    //get value online status
                    String onlineStatus = ""+ ds.child("onlineStatus").getValue();
                    if (onlineStatus.equals("online")){
                        userStatusTv.setText(onlineStatus);
                    } else {
                        // convert time stamp to proper time date
//                        Calendar cal = Calendar.getInstance(Locale.ENGLISH);
//                        cal.setTimeInMillis(Long.parseLong(onlineStatus));
//                        String datetime = DateFormat.format("dd/MM/yyyy hh:mm aa",cal).toString();
//                        userStatusTv.setText("Last seen at: "+ datetime);

                        GetTimeAgo getTimeAgo = new GetTimeAgo();
                        long lastTime = Long.parseLong(onlineStatus);
                        String lastSeenTime = getTimeAgo.getTimeAgo(lastTime, getApplicationContext());

                        userStatusTv.setText("last seen "+ lastSeenTime);
                    }

                    //set data
                    nameTv.setText(name);
                    if (!hisImage.equals("")) {
                        try {
//                            Picasso.get().load(hisImage).placeholder(R.drawable.ic_default_img_white).into(profileIv);
                            Glide.with(getApplicationContext())
                                    .load(hisImage)
                                    .placeholder(R.drawable.ic_default_img_white)
                                    .into(profileIv);
                        }catch (Exception e){
//                            Picasso.get().load(R.drawable.ic_default_img_white).into(profileIv);
                            Glide.with(getApplicationContext())
                                    .load(R.drawable.ic_default_img_white)
                                    .into(profileIv);
                        }
                    }
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });

        // click button to send message
        sendBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                notify = true;
                //get text from edit text
                String message = messageEt.getText().toString().trim();
                //check if text is empty or not
                if(TextUtils.isEmpty(message)){
                    //text empty
                    Toast.makeText(ChatActivity.this, "Cannot send empty message...", Toast.LENGTH_SHORT).show();
                } else{
                    //text not empty
                    sendMessage(message);
                }
                //reset edit text after sending message
                messageEt.setText("");
            }
        });

        readMessage();

        seenMessage();

    }

    private void seenMessage() {
        userRefForSeen = FirebaseDatabase.getInstance().getReference("Chats");
        seenListener = userRefForSeen.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                for (DataSnapshot ds: snapshot.getChildren()){
                    ModelChat chat = ds.getValue(ModelChat.class);
                    if (chat.getReceiver().equals(myUid) && chat.getSender().equals(hisUid)){
                        HashMap<String,Object> hasSeenMap = new HashMap<>();
                        hasSeenMap.put("isSeen",true);
                        ds.getRef().updateChildren(hasSeenMap);
                    }
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });
    }

    private void readMessage() {
        chatList = new ArrayList<>();
        DatabaseReference dbRef = FirebaseDatabase.getInstance().getReference("Chats");
        dbRef.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                chatList.clear();
                for (DataSnapshot ds: snapshot.getChildren()){
                    ModelChat chat = ds.getValue(ModelChat.class);
                    if (chat.getReceiver().equals(myUid) && chat.getSender().equals(hisUid) ||
                            chat.getReceiver().equals(hisUid) && chat.getSender().equals(myUid) ){
                        chatList.add(chat);
                    }
                    // adapter
                    adapterChat = new AdapterChat(ChatActivity.this, chatList, hisImage);
                    adapterChat.notifyDataSetChanged();
                    //set adapter to recyclerview
                    recyclerView.setAdapter(adapterChat);
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });
    }

    private void sendMessage(String message) {
        DatabaseReference databaseReference = FirebaseDatabase.getInstance().getReference();

        String timestamp = String.valueOf(System.currentTimeMillis());

        HashMap<String, Object> hashMap = new HashMap<>();
        hashMap.put("sender", myUid);
        hashMap.put("receiver", hisUid);
        hashMap.put("message", message);
        hashMap.put("timestamp", timestamp);
        hashMap.put("isSeen", false);
        databaseReference.child("Chats").push().setValue(hashMap);

        DatabaseReference database = FirebaseDatabase.getInstance().getReference("Users").child(myUid);
        database.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                ModelUser user = snapshot.getValue(ModelUser.class);

                if (notify) {
                    sendNotification(hisUid, user.getName(), message);
                }
                notify = false;
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });

        //create chatList node/child in firebase database
        DatabaseReference chatRef1 = FirebaseDatabase.getInstance().getReference("Chatlist")
                .child(myUid)
                .child(hisUid);
        chatRef1.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                if (!snapshot.exists()){
                    chatRef1.child("id").setValue(hisUid);
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });
        DatabaseReference chatRef2 = FirebaseDatabase.getInstance().getReference("Chatlist")
                .child(hisUid)
                .child(myUid);
        chatRef2.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                if (!snapshot.exists()){
                    chatRef2.child("id").setValue(myUid);
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });
    }

    private void sendNotification(final String hisUid, final String name, final String message) {
        DatabaseReference allTokens = FirebaseDatabase.getInstance().getReference("Tokens");
        Query query = allTokens.orderByKey().equalTo(hisUid);
        query.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                for (DataSnapshot ds: snapshot.getChildren()){
                    Token token = ds.getValue(Token.class);
                    Data data = new Data(myUid, name+": "+message, "New Message", hisUid, "", R.drawable.ic_default_img);

                    Sender sender = new Sender(data, token.getToken());
                    apiService.sendNotification(sender)
                            .enqueue(new Callback<Response>() {
                                @Override
                                public void onResponse(Call<Response> call, retrofit2.Response<Response> response) {
//                                    Toast.makeText(ChatActivity.this, ""+response.message(), Toast.LENGTH_SHORT).show();
                                }

                                @Override
                                public void onFailure(Call<Response> call, Throwable t) {

                                }
                            });
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });
    }

    private void checkUserStatus(){
        //get current user
        FirebaseUser user = firebaseAuth.getCurrentUser();
        if(user != null){
            //user is signed in stay here
            //set email of logged in user
            //mProfileTv.setText(user.getEmail());
            myUid = user.getUid(); //currently sign in user's
        }else{
            //user not signed in, go to main activity
            startActivity(new Intent(this, MainActivity.class));
            finish();
        }
    }

    private void checkOnlineStatus(String status) {
        DatabaseReference dbRef = FirebaseDatabase.getInstance().getReference("Users").child(myUid);
        HashMap<String, Object> hashMap = new HashMap<>();
        hashMap.put("onlineStatus", status);
        //update value of online status of current user
        dbRef.updateChildren(hashMap);
    }


    @Override
    protected void onStart() {
        checkUserStatus();
        //set online
        checkOnlineStatus("online");
        super.onStart();
    }

    @Override
    protected void onPause() {
        super.onPause();
        String timestamp = String.valueOf(System.currentTimeMillis());

        //set offline with last seen timestamp
        checkOnlineStatus(timestamp);
        userRefForSeen.removeEventListener(seenListener);
    }

    @Override
    protected void onResume() {
        //set online
        checkOnlineStatus("online");
        super.onResume();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        //hide search view
        menu.findItem(R.id.action_search).setVisible(false);
        menu.findItem(R.id.action_nearby_clinic).setVisible(false);
        menu.findItem(R.id.action_chatlist).setVisible(false);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        //get item id
        int id = item.getItemId();
        if (id == R.id.action_logout){
            String timestamp = String.valueOf(System.currentTimeMillis());

            //set offline with last seen timestamp
            checkOnlineStatus(timestamp);
            firebaseAuth.signOut();
            checkUserStatus();
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return true;
    }
}

File: Healthia/app/src/main/java/com/example/healthia/ChatActivity.kt
package com.example.healthia

import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.text.format.DateFormat
import android.view.*
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.healthia.adapters.AdapterChat
import com.example.healthia.models.ModelChat
import com.example.healthia.models.ModelUser
import com.example.healthia.notifications.*
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*
import com.squareup.picasso.Picasso
import retrofit2.Call
import retrofit2.Callback
import java.util.*

class ChatActivity : AppCompatActivity() {
    //views from xml
    var toolbar: Toolbar? = null
    var recyclerView: RecyclerView? = null
    var profileIv: ImageView? = null
    var nameTv: TextView? = null
    var userStatusTv: TextView? = null
    var messageEt: EditText? = null
    var sendBtn: ImageButton? = null

    //firebase auth
    var firebaseAuth: FirebaseAuth? = null
    var firebaseDatabase: FirebaseDatabase? = null
    var usersDbRef: DatabaseReference? = null

    //for checking if use has seen message or not
    var seenListener: ValueEventListener? = null
    var userRefForSeen: DatabaseReference? = null
    var chatList: MutableList<ModelChat?>? = null
    var adapterChat: AdapterChat? = null
    var hisUid: String? = null
    var myUid: String? = null
    var hisImage: String? = null
    var apiService: APIService? = null
    var notify = false
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_chat)

        //init views
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        toolbar.title = ""
        recyclerView = findViewById(R.id.chat_recyclerView)
        profileIv = findViewById(R.id.profileIv)
        nameTv = findViewById(R.id.nameTv)
        userStatusTv = findViewById(R.id.userStatusTv)
        messageEt = findViewById(R.id.messageEt)
        sendBtn = findViewById(R.id.sendBtn)

        //layout (linearlayout) for recyclerview
        val linearLayoutManager = LinearLayoutManager(this)
        linearLayoutManager.stackFromEnd = true
        //recyclerview properties
        recyclerView?.setHasFixedSize(true)
        recyclerView?.layoutManager = linearLayoutManager

        //create api service
        apiService = Client.getRetrofit("https://fcm.googleapis.com/")!!.create(APIService::class.java)
        val intent = intent
        hisUid = intent.getStringExtra("hisUid")

        //firebase auth instance
        firebaseAuth = FirebaseAuth.getInstance()
        firebaseDatabase = FirebaseDatabase.getInstance()
        usersDbRef = firebaseDatabase!!.getReference("Users")

        //search user to get user's info
        val userQuery = usersDbRef!!.orderByChild("uid").equalTo(hisUid)

        //get user's picture and name
        userQuery.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                //check until required info is received
                for (ds in snapshot.children) {
                    //get data
                    val name = "" + ds.child("name").value
                    hisImage = "" + ds.child("image").value
                    //get value online status
                    val onlineStatus = "" + ds.child("onlineStatus").value
                    if (onlineStatus == "online") {
                        userStatusTv?.text = onlineStatus
                    } else {
                        // convert time stamp to proper time date
                        val cal = Calendar.getInstance(Locale.ENGLISH)
                        cal.timeInMillis = onlineStatus.toLong()
                        val datetime = DateFormat.format("dd/MM/yyyy hh:mm aa", cal).toString()
                        userStatusTv?.text = "Last seen at: $datetime"
                    }

                    //set data
                    nameTv?.text = name
                    try {
                        Picasso.get().load(hisImage).placeholder(R.drawable.ic_default_img_white).into(profileIv)
                    } catch (e: Exception) {
                        Picasso.get().load(R.drawable.ic_default_img_white).into(profileIv)
                    }
                }
            }

            override fun onCancelled(error: DatabaseError) {}
        })

        // click button to send message
        sendBtn?.setOnClickListener(View.OnClickListener {
            notify = true
            //get text from edit text
            val message = messageEt?.text.toString().trim { it <= ' ' }
            //check if text is empty or not
            if (TextUtils.isEmpty(message)) {
                //text empty
                Toast.makeText(this@ChatActivity, "Cannot send empty message...", Toast.LENGTH_SHORT).show()
            } else {
                //text not empty
                sendMessage(message)
            }
            //reset edit text after sending message
            messageEt?.setText("")
        })
        readMessage()
        seenMessage()
    }

    private fun seenMessage() {
        userRefForSeen = FirebaseDatabase.getInstance().getReference("Chats")
        seenListener = userRefForSeen!!.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                for (ds in snapshot.children) {
                    val chat = ds.getValue(ModelChat::class.java)
                    if (chat?.receiver == myUid && chat?.sender == hisUid) {
                        val hasSeenMap = HashMap<String, Any>()
                        hasSeenMap["isSeen"] = true
                        ds.ref.updateChildren(hasSeenMap)
                    }
                }
            }

            override fun onCancelled(error: DatabaseError) {}
        })
    }

    private fun readMessage() {
        chatList = ArrayList()
        val dbRef = FirebaseDatabase.getInstance().getReference("Chats")
        dbRef.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                (chatList as ArrayList<ModelChat?>).clear()
                for (ds in snapshot.children) {
                    val chat = ds.getValue(ModelChat::class.java)
                    if (chat?.receiver == myUid && chat?.sender == hisUid ||
                            chat?.receiver == hisUid && chat?.sender == myUid) {
                        (chatList as ArrayList<ModelChat?>).add(chat)
                    }
                    // adapter
                    adapterChat = AdapterChat(this@ChatActivity, chatList as ArrayList<ModelChat?>, hisImage)
                    adapterChat!!.notifyDataSetChanged()
                    //set adapter to recyclerview
                    recyclerView!!.adapter = adapterChat
                }
            }

            override fun onCancelled(error: DatabaseError) {}
        })
    }

    private fun sendMessage(message: String) {
        val databaseReference = FirebaseDatabase.getInstance().reference
        val timestamp = System.currentTimeMillis().toString()
        val hashMap = HashMap<String, Any?>()
        hashMap["sender"] = myUid
        hashMap["receiver"] = hisUid
        hashMap["message"] = message
        hashMap["timestamp"] = timestamp
        hashMap["isSeen"] = false
        databaseReference.child("Chats").push().setValue(hashMap)
        val database = FirebaseDatabase.getInstance().getReference("Users").child(myUid!!)
        database.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                val user = snapshot.getValue(ModelUser::class.java)
                if (notify) {
                    sendNotification(hisUid, user?.name, message)
                }
                notify = false
            }

            override fun onCancelled(error: DatabaseError) {}
        })

        //create chatList node/child in firebase database
        val chatRef1 = FirebaseDatabase.getInstance().getReference("Chatlist")
                .child(myUid!!)
                .child(hisUid!!)
        chatRef1.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                if (!snapshot.exists()) {
                    chatRef1.child("id").setValue(hisUid)
                }
            }

            override fun onCancelled(error: DatabaseError) {}
        })
        val chatRef2 = FirebaseDatabase.getInstance().getReference("Chatlist")
                .child(hisUid!!)
                .child(myUid!!)
        chatRef2.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                if (!snapshot.exists()) {
                    chatRef2.child("id").setValue(myUid)
                }
            }

            override fun onCancelled(error: DatabaseError) {}
        })
    }

    private fun sendNotification(hisUid: String?, name: String?, message: String) {
        val allTokens = FirebaseDatabase.getInstance().getReference("Tokens")
        val query = allTokens.orderByKey().equalTo(hisUid)
        query.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                for (ds in snapshot.children) {
                    val token = ds.getValue(Token::class.java)
                    val data = Data(myUid, "$name: $message", "New Message", hisUid, R.drawable.ic_default_img)
                    val sender = Sender(data, token?.token)
                    apiService!!.sendNotification(sender)
                            .enqueue(object : Callback<Response?> {
                                override fun onResponse(call: Call<Response?>, response: retrofit2.Response<Response?>) {
                                    Toast.makeText(this@ChatActivity, "" + response.message(), Toast.LENGTH_SHORT).show()
                                }

                                override fun onFailure(call: Call<Response?>, t: Throwable) {}
                            })
                }
            }

            override fun onCancelled(error: DatabaseError) {}
        })
    }

    private fun checkUserStatus() {
        //get current user
        val user = firebaseAuth!!.currentUser
        if (user != null) {
            //user is signed in stay here
            //set email of logged in user
            //mProfileTv.setText(user.getEmail());
            myUid = user.uid //currently sign in user's
        } else {
            //user not signed in, go to main activity
            startActivity(Intent(this, LoginOrRegisterActivity::class.java))
            finish()
        }
    }

    private fun checkOnlineStatus(status: String) {
        val dbRef = FirebaseDatabase.getInstance().getReference("Users").child(myUid!!)
        val hashMap = HashMap<String, Any>()
        hashMap["onlineStatus"] = status
        //update value of online status of current user
        dbRef.updateChildren(hashMap)
    }

    override fun onStart() {
        checkUserStatus()
        //set online
        checkOnlineStatus("online")
        super.onStart()
    }

    override fun onPause() {
        super.onPause()
        val timestamp = System.currentTimeMillis().toString()

        //set offline with last seen timestamp
        checkOnlineStatus(timestamp)
        userRefForSeen!!.removeEventListener(seenListener!!)
    }

    override fun onResume() {
        //set online
        checkOnlineStatus("online")
        super.onResume()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        //hide search view
        menu.findItem(R.id.action_search).isVisible = false
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        //get item id
        val id = item.itemId
        if (id == R.id.action_logout) {
            val timestamp = System.currentTimeMillis().toString()

            //set offline with last seen timestamp
            checkOnlineStatus(timestamp)
            firebaseAuth!!.signOut()
            checkUserStatus()
        }
        return super.onOptionsItemSelected(item)
    }
}